### PR TITLE
[serve] Fix serve agent 501 response

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

TL;DR: the optimization added in https://github.com/ray-project/ray/pull/38836, which returns 501 if Serve dependencies aren't installed, doesn't work. Instead of returning a 501 response like it should, the serve agent returns a mysterious `TypeError: object Response can't be used in 'await' expression`

The decorator `gracefully_handle_missing_serve_dependencies` doesn't return an async function, and the second decorator `init_ray_and_catch_exceptions` runs `await f`, `f` being the result of the first decorator.
- Normally this works fine since `f` returns a coroutine (the original async function)
- But when serve dependencies are not installed and `f` returns a 501 Response, await on the 501 Response fails.

The tests `test_dashboard_does_not_depend_on_serve` and `test_agent_does_not_depend_on_serve` didn't catch this because it includes the assert statement inside a try-except block, so the assert statement was actually failing but it was being ignored by the except block :facepalm:

Fixed the test and also verified manually this time that the optimization works.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/tests/test_dashboard.py [hunk 1]
```diff
@@ -925,11 +925,12 @@ def test_dashboard_does_not_depend_on_serve():
     # Check that Serve-dependent features fail
     try:
         response = requests.get(f"http://{agent_url}/api/serve/deployments/")
-        assert response.status_code == 501
     except Exception as e:
         # Fail to connect to service is fine.
         print(e)
-        assert True
+    else:
```
Comment (edoakes) [inline dashboard/tests/test_dashboard.py:931]: wow, I did not know that you could use an `else` block after an exception handler until today... this is tricky and seems like it might confuse people in the future. can you rewrite this in a way to avoid this syntax?

in general we shouldn't be catching bare `Exception` types because it causes subtle bugs like the one you uncovered here. maybe catch a more specific error in that block and then keep the assert in the `try` block?

File: dashboard/tests/test_dashboard.py [hunk 2]
```diff
@@ -896,16 +896,16 @@ def test_gcs_check_alive(fast_gcs_failure_detection, ray_start_with_dashboard):
     assert dashboard_proc.wait(10) == 255
 
 
-@pytest.mark.skipif(
-    os.environ.get("RAY_DEFAULT") != "1",
-    reason="This test only works for default installation.",
-)
+# @pytest.mark.skipif(
```
Comment (shrekris-anyscale) [inline dashboard/tests/test_dashboard.py:899]: Make sure to remove the commented code when you no longer need it.

File: dashboard/tests/test_dashboard.py (hunk 1 above)
Comment (zcin) [reply dashboard/tests/test_dashboard.py]: Switched to catching `requests.ConnectionError`!

File: dashboard/tests/test_dashboard.py [hunk 3]
```diff
@@ -904,8 +904,8 @@ def test_dashboard_does_not_depend_on_serve():
     """Check that the dashboard can start without Serve."""
     ray.shutdown()
 
-    with pytest.raises(ImportError):
-        from ray import serve  # noqa: F401
+    # with pytest.raises(ImportError):
+    #     from ray import serve  # noqa: F401
```
Comment (edoakes) [inline dashboard/tests/test_dashboard.py:908]: is this intentional?

**Review by sihanwang41 — APPROVED** [review]
thx for catching it.

**Comment by zcin** [comment]
@edoakes This change is ready to merge!
